asfadmin/thin-egress-app
# Add Open Telemetry configuration and tracing

This is a redo of a previous PR. The intention here is to clean up a batch of duplicate commits. The original PR with discussion: https://github.com/asfadmin/thin-egress-app/pull/448

Includes
1. CloudFormation changes to add the AWS Open Telemetry Instrumentation Layer
2. Python changes to set spans in the Lambda
3. Collector configuration for exporting traces
4. Makefile changes to include the collector configuration

### Testing:

This should be tested thoroughly with TEA's build pipeline. To test:

1. Build TEA as normal. `app.py` now includes imports from the opentelemetry package from the Lambda Layer. This will build normally but should be noted as it would prevent any local running of the code.

2.Deploy TEA with the variable `OTEL_TRACES_SAMPLER` set to `always_on` (or change this variable after deployment).

3. Add an example file to the protected bucket.

4. Try downloading that file using TEA's GET endpoint. e.g. `https://yvr8cds2g3.execute-api.us-east-1.amazonaws.com:7000/DEV/ltt-test-protected/MOD09GQ.A3864382.mN5egI.006.9435156614520.hdf`

5. Check the CloudWatch logs for spans and the Trace Analytics plugin in the appropriate Cloud Metrics environment for traces. Spans in the CloudWatch logs would look like:

```
Span #1 
Trace ID: 456
Parent ID: 123
```

The Cloud Metrics environment is set with the `APM_SERVER_URL` ENV. The default right now is SIT.
